### ISSUE TITLE
`pip install h5py` tries to install an old version of numpy

### ISSUE THREAD

**Issue by dstansby** [h5py/h5py#1981]
In a clean conda environment, running the commands below fails, because `h5py` tries to install numpy 1.19.3 which is not compatible with python 3.10. I'm not sure why this is the case - this block implies that `h5py` should on enforce a minimum version, instead of installing an exact version of numpy https://github.com/h5py/h5py/blob/fb9989a700e60cd1c1dec20e329218d449e039cf/setup.py#L28-L39

```python
conda install python=3.10
pip install numpy # Installs numpy==1.21.3
pip install h5py
```
Output of final command:
```python
$ pip install h5py --verbose
Using pip 21.2.4 from /Users/dstansby/miniconda3/envs/py310/lib/python3.10/site-packages/pip (python 3.10)
Collecting h5py
  Using cached h5py-3.5.0.tar.gz (384 kB)
  Running command /Users/dstansby/miniconda3/envs/py310/bin/python /private/var/folders/3j/f7vtq4q1397bxjn19wftn_fm0000gn/T/pip-standalone-pip-nestuxec/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /private/var/folders/3j/f7vtq4q1397bxjn19wftn_fm0000gn/T/pip-build-env-3u2eo2pe/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- setuptools wheel
  Collecting setuptools
    Using cached setuptools-58.2.0-py3-none-any.whl (946 kB)
  Collecting wheel
    Using cached wheel-0.37.0-py2.py3-none-any.whl (35 kB)
  Installing collected packages: wheel, setuptools
  Successfully installed setuptools-58.2.0 wheel-0.37.0
  Installing build dependencies ... done
  Running command /Users/dstansby/miniconda3/envs/py310/bin/python /Users/dstansby/miniconda3/envs/py310/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py get_requires_for_build_wheel /var/folders/3j/f7vtq4q1397bxjn19wftn_fm0000gn/T/tmp9i7s2cwc
  Getting requirements to build wheel ... done
  Running command /Users/dstansby/miniconda3/envs/py310/bin/python /private/var/folders/3j/f7vtq4q1397bxjn19wftn_fm0000gn/T/pip-standalone-pip-b06wwdt6/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /private/var/folders/3j/f7vtq4q1397bxjn19wftn_fm0000gn/T/pip-build-env-3u2eo2pe/normal --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'numpy==1.17.5; python_version == "3.8"' 'Cython>=0.29.15; python_version >= "3.9"' pkgconfig 'numpy==1.19.3; python_version >= "3.9"' 'numpy==1.14.5; python_version == "3.7"' 'Cython>=0.29; python_version < "3.8"' 'Cython>=0.29.14; python_version == "3.8"'
  Ignoring numpy: markers 'python_version == "3.8"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.7"' don't match your environment
  Ignoring Cython: markers 'python_version < "3.8"' don't match your environment
  Ignoring Cython: markers 'python_version == "3.8"' don't match your environment
  Collecting Cython>=0.29.15
    Using cached Cython-0.29.24-py2.py3-none-any.whl (979 kB)
  Collecting pkgconfig
    Using cached pkgconfig-1.5.5-py3-none-any.whl (6.7 kB)
  Collecting numpy==1.19.3
    Using cached numpy-1.19.3.zip (7.3 MB)
    Installing build dependencies: started
^C  Installing backend dependencies ... canceled
```

**Comment by tacaswell** [comment]
And the way `NUMPY_MIN_VERSIONS` works is that it sets the version of numpy the wheel is built with which ensures that it will work with any _newer_ version of numpy due to the api/abi guarantees that numpy gives us.  The above is an escape hatch to get out of the build system building you a safe wheel.

**Comment by henryiii** [comment]
Not funded yet, but there's also https://iscinumpy.gitlab.io/post/scikit-build-proposal/. CMake is much more widely used than meson in the scientific community, so it's focus is for packages wanting to leverage that, rather than rebuilding the build system from scratch.

> I think this means that your wheels claim compatibility with a broader range of numpy versions than they'll actually work with. I guess that doesn't often matter in practice, but if I've got a way to get the metadata right...

The point of `oldest-supported-numpy` is that it picks the oldest version of numpy available for a certain platform and builds with that. Then you just set your actual design minimum for the runtime - you can't get an older version of numpy because there isn't one for that platform. For example, say you set `numpy<=1.17` for runtime. Then on Python 3.10, you'll never get an older version of numpy that what's supported on that platform (1.21, IIRC), and that's what oldest-supported-numpy built with. The fact your runtime allows numpy 1.17-20 doesn't matter; those aren't available for that system. You don't work with it, sure - but numpy itself won't work anyway.

**Comment by takluyver** [comment]
This should be resolved by version 3.6, which I just released. Resolved both in that there are now wheels for Python 3.10, and we're using oldest-supported-numpy, which will hopefully avoid this particular problem on future Python releases. (It will still try to build h5py from source until there's a wheel available, which will fail in a lot of cases)
